packages/protocol/contracts/governance/Governance.sol

```diff
@@ -615,30 +653,66 @@ contract Governance is
    * @notice Approves the hash of a hotfix transaction(s).
    * @param hash The abi encoded keccak256 hash of the hotfix transaction(s) to be approved.
    */
-  function approveHotfix(bytes32 hash) external hotfixNotExecuted(hash) onlyApprover {
-    hotfixes[hash].approved = true;
-    emit HotfixApproved(hash);
+  function approveHotfix(
+    bytes32 hash
+  ) external hotfixNotExecuted(hash) onlyApproverOrSecurityCouncil {
+    if (msg.sender == approver) {
+      hotfixes[hash].approved = true;
+    } else {
+      if (isL2()) {
+        hotfixes[hash].councilApproved = true;
+      } else {
+        revert("Hotfix approval by security council is not available on L1.");
+      }
+    }
+    emit HotfixApproved(hash, msg.sender);
   }
 
   /**
    * @notice Whitelists the hash of a hotfix transaction(s).
    * @param hash The abi encoded keccak256 hash of the hotfix transaction(s) to be whitelisted.
    */
-  function whitelistHotfix(bytes32 hash) external hotfixNotExecuted(hash) {
-    hotfixes[hash].whitelisted[msg.sender] = true;
+  function whitelistHotfix(bytes32 hash) external hotfixNotExecuted(hash) onlyL1 {
+    hotfixes[hash].deprecated_whitelisted[msg.sender] = true;
     emit HotfixWhitelisted(hash, msg.sender);
   }
 
   /**
-   * @notice Gives hotfix a prepared epoch for execution.
+   * @notice Gives hotfix a time limit for execution.
    * @param hash The hash of the hotfix to be prepared.
    */
   function prepareHotfix(bytes32 hash) external hotfixNotExecuted(hash) {
-    require(isHotfixPassing(hash), "hotfix not whitelisted by 2f+1 validators");
-    uint256 epoch = getEpochNumber();
-    require(hotfixes[hash].preparedEpoch < epoch, "hotfix already prepared for this epoch");
-    hotfixes[hash].preparedEpoch = epoch;
-    emit HotfixPrepared(hash, epoch);
+    if (isL2()) {
+      uint256 _currentTime = now;
+      require(hotfixExecutionTimeWindow > 0, "Hotfix execution time window not set");
+      require(
+        hotfixes[hash].executionTimeLimit < _currentTime,
+        "Hotfix already prepared for this timeframe."
+      );
+      require(hotfixes[hash].approved, "Hotfix not approved by approvers.");
+      require(hotfixes[hash].councilApproved, "Hotfix not approved by security council.");
+
+      if (
```

Comment:
What is the reasoning behind resetting hotfix ?

Also it should probably be new function

Reply:
I explained [here](https://github.com/celo-org/celo-monorepo/pull/11014#issuecomment-2146175966). But the gist is that it prevents a hotfix from being executed long after it was approved and requires re-approval if too much time elapsed since it was last prepared.

Reply:
I extracted it to an internal function, but I am relunctant to make it a seperate step, as `prepareHotfix()` should not be allowed to prepare a timed-out hotfix.